### PR TITLE
[OpenXLA] Pull xla from openxla and using cc_binary and rebased with Apr 23

### PR DESCRIPTION
**Pull request by ManfeiBai**

### CODE REVIEW

**Comment by ManfeiBai** [comment]
GPU test failure first because of `"@org_tensorflow//tensorflow/compiler/jit:xla_gpu_device",`, according to https://github.com/pytorch/xla/pull/4929/files

**Comment by ManfeiBai** [comment]
old gpu failure: https://app.circleci.com/pipelines/github/pytorch/xla/18654/workflows/0f4fa3fd-ec54-42c7-b114-2816e425789f/jobs/45671

File: third_party/xla_client/BUILD [hunk 1]
```diff
@@ -77,70 +95,72 @@ cc_library(
         ":unique",
         ":util",
         ":xla_util",
-        ":xrt_local_service",
-        ":xrt_session",
-        ":xrt_session_cache",
-        "@org_tensorflow//tensorflow:grpc++",
-        "@org_tensorflow//tensorflow/cc:client_session",
-        "@org_tensorflow//tensorflow/cc:scope",
-        "@org_tensorflow//tensorflow/compiler/jit:xla_cpu_device",
```
Comment (miladm) [inline third_party/xla_client/BUILD:86]: looks like none of `jit` lines are support. can you confirm this line is supported @ManfeiBai?

cc @JackCaoG

File: third_party/xla_client/BUILD [hunk 2]
```diff
@@ -77,70 +95,72 @@ cc_library(
...
         "@com_google_absl//absl/numeric:int128",
         "@com_google_absl//absl/strings",
     ] + if_cuda_is_configured([
-        "@org_tensorflow//tensorflow/compiler/jit:xla_gpu_device",
-        "@org_tensorflow//tensorflow/compiler/xla/stream_executor:cuda_platform",
+        # "@org_tensorflow//tensorflow/compiler/jit:xla_gpu_device",
+        # "@xla//xla/service:gpu_plugin",
+        "@xla//xla/stream_executor:cuda_platform",
     ]) + if_with_tpu_support([
-        "@org_tensorflow//tensorflow/compiler/jit:xla_tpu_device",
```
Comment (miladm) [inline third_party/xla_client/BUILD:142]: looks like we have the `jit` issue with this line also. please confirm.

File: third_party/xla_client/BUILD [hunk 3]
```diff
@@ -77,70 +95,72 @@ cc_library(
...
-        "@org_tensorflow//tensorflow/compiler/jit:xla_tpu_jit",
```
Comment (miladm) [inline third_party/xla_client/BUILD:143]: ditto

File: torch_xla/csrc/convolution.cpp [hunk 4]
```diff
@@ -1,16 +1,34 @@
 #include "torch_xla/csrc/convolution.h"
 
-#include "tensorflow/compiler/tf2xla/kernels/conv_op_helpers.h"
-#include "tensorflow/compiler/xla/client/lib/constants.h"
-#include "tensorflow/core/framework/tensor_shape.h"
-#include "tensorflow/core/kernels/conv_grad_ops.h"
+#include "third_party/xla_client/conv_op_helpers.h"
+#include "xla/client/lib/constants.h"
+// #include "tensorflow/core/framework/tensor_shape.h"
+// #include "tensorflow/core/kernels/conv_grad_ops.h"
```
Comment (miladm) [inline torch_xla/csrc/convolution.cpp:6]: What's the replacement for this line in openxla @ManfeiBai?

File: torch_xla/csrc/convolution.cpp [hunk 5]
```diff
@@ -1,16 +1,34 @@
 #include "torch_xla/csrc/convolution.h"
 
-#include "tensorflow/compiler/tf2xla/kernels/conv_op_helpers.h"
-#include "tensorflow/compiler/xla/client/lib/constants.h"
-#include "tensorflow/core/framework/tensor_shape.h"
-#include "tensorflow/core/kernels/conv_grad_ops.h"
+#include "third_party/xla_client/conv_op_helpers.h"
+#include "xla/client/lib/constants.h"
+// #include "tensorflow/core/framework/tensor_shape.h"
```
Comment (miladm) [inline torch_xla/csrc/convolution.cpp:5]: What's the replacement for this line in openxla @ManfeiBai?

File: third_party/xla_client/topology.proto [hunk 6]
```diff
@@ -0,0 +1,49 @@
+syntax = "proto3";
```
Comment (miladm) [inline third_party/xla_client/topology.proto:1]: I don't think we should automatically decide to move topology.proto to torch_ala. Is this for verification purposes?

File: third_party/xla_client/pjrt_computation_client.cc [hunk 7]
```diff
@@ -529,7 +529,7 @@ PjRtComputationClient::ExecuteReplicated(
   execute_options.multi_slice_config = nullptr;
 
   // Required as of cl/518733871
-  execute_options.use_major_to_minor_data_layout_for_callbacks = true;
+  // execute_options.use_major_to_minor_data_layout_for_callbacks = true;
```
Comment (miladm) [inline third_party/xla_client/pjrt_computation_client.cc:532]: why is this commented out?

File: WORKSPACE [hunk 8]
```diff
@@ -1,32 +1,31 @@
...
-    ],
-    strip_prefix = "tensorflow-5e229cfaa3c13c389f864b0ed38877fb61936161",
+    name = "xla",
+    # patch_args = [
+    #     "-l",
+    #     "-p1",
+    # ],
+    # patch_tool = "patch",
+    # patches = [
+    #     "//openxla_patches:cache_urls.diff",
```
Comment (miladm) [inline WORKSPACE:16]: why is this section commented out?

File: torch_xla/csrc/convolution.cpp [hunk 9]
```diff
@@ -144,15 +162,17 @@ tensorflow::ConvOpAttrs MakeConvOpAttrs(
...
   // explicit_padding requires to have (spatial_dims + 2) * 2 elements
   conv_op_attrs.explicit_paddings.resize(4);
   for (int spatial_dim = 0; spatial_dim < num_spatial_dims; ++spatial_dim) {
     conv_op_attrs.explicit_paddings.push_back(spatial_padding[spatial_dim]);
     conv_op_attrs.explicit_paddings.push_back(spatial_padding[spatial_dim]);
   }
-  conv_op_attrs.data_format = tensorflow::TensorFormat::FORMAT_NCHW;
+  conv_op_attrs.data_format = MakeConvolutionDimensionNumbers(
+      xla::XLATensorFormat::FORMAT_NCHW, num_spatial_dims);
+  // xla::XLATensorFormat::FORMAT_NCHW; // 1; //tensorflow::TensorFormat::FORMAT_NCHW; // use `ConvolutionDimensionNumbers` from /pytorch/tensorflow/tensorflow/compiler/xla/xla_data.proto ?
```
Comment (miladm) [inline torch_xla/csrc/convolution.cpp:175]: what's this line? why is it commented?

File: torch_xla/csrc/init_python_bindings.cpp [hunk 10]
```diff
@@ -19,15 +19,16 @@
...
-#include "tensorflow/compiler/xla/service/hlo_parser.h"
-#include "tensorflow/core/example/example.pb.h"
-#include "tensorflow/core/example/feature.pb.h"
-#include "tensorflow/python/profiler/internal/profiler_pywrap_impl.h"
-#include "tensorflow/tsl/platform/env.h"
-#include "tensorflow/tsl/profiler/lib/traceme.h"
+#include "xla/pjrt/distributed/distributed.h"
+#include "xla/python/profiler/internal/traceme_wrapper.h"
+#include "xla/service/hlo_parser.h"
+#include "third_party/xla_client/example.pb.h"
```
Comment (miladm) [inline torch_xla/csrc/init_python_bindings.cpp:25]: why is this file moved to `xla_client`? is it for prototyping purposes?

CC @JackCaoG @ManfeiBai

File: torch_xla/csrc/convolution.cpp (hunk 4 above)
Comment (ManfeiBai) [reply torch_xla/csrc/convolution.cpp:6]: Thanks, @miladm ,

not found replacement from OpenXLA

File: torch_xla/csrc/init_python_bindings.cpp (hunk 10 above)
Comment (ManfeiBai) [reply torch_xla/csrc/init_python_bindings.cpp:25]: yes, it is for prototyping purposes,
but I created a `example.pb.h` in `third_party/xla_client` directly because want to only add new files in `third_party/xla_client`

File: torch_xla/csrc/convolution.cpp (hunk 9 above)
Comment (ManfeiBai) [reply torch_xla/csrc/convolution.cpp:175]: this line is my comments and saved my logic history in case I need to recall it in the future

File: WORKSPACE (hunk 8 above)
Comment (ManfeiBai) [reply WORKSPACE:16]: I didn't patch PTXLA with patches, because I didn't check patches, but we build successfully, so I skipped this patch part and commented out it

File: third_party/xla_client/topology.proto (hunk 6 above)
Comment (ManfeiBai) [reply third_party/xla_client/topology.proto:1]: copied this line to bypass this problem, and this line is used to define the syntax is `proto3`, because format requirement;

not for verification purpose;

File: setup.py
Comment (ManfeiBai) [inline setup.py:1]: this part has been deleted in the newest code, so skip

File: test/cpp/CMakeLists.txt
Comment (ManfeiBai) [inline test/cpp/CMakeLists.txt:1]: this file disappear in the newest code, so will skip this code

File: test/pjrt/test_experimental_pjrt.py
Comment (ManfeiBai) [inline test/pjrt/test_experimental_pjrt.py:1]: this file has been renamed in https://github.com/pytorch/xla/pull/5011/files

File: test/run_tests.sh [hunk 11]
```diff
@@ -169,7 +161,7 @@ function run_xla_op_tests {
   run_test_without_functionalization "$CDIR/test_operations.py" "$@" --verbosity=$VERBOSITY
   run_test "$CDIR/test_async_closures.py"
   run_test "$CDIR/test_xla_dist.py"
-  run_test "$CDIR/test_profiler.py"
+  # run_test "$CDIR/test_profiler.py"
   run_test "$CDIR/test_ops.py"
```
Comment (ManfeiBai) [inline test/run_tests.sh:165]: this has been fixed to tsl

**Comment by ManfeiBai** [comment]
this PR has been a very important PR or milestone, work continued in https://github.com/pytorch/xla/pull/5202
